File: apps/ios/ConstructionManager/ConstructionManager/Models/Report.swift
//
//  Report.swift
//  ConstructionManager
//
//  Reports and Analytics models
//

import Foundation

// MARK: - Report Type
enum ReportType: String, Codable, CaseIterable {
    case labor = "LABOR"
    case equipment = "EQUIPMENT"
    case safety = "SAFETY"
    case safetyMeetings = "SAFETY_MEETINGS"
    case financial = "FINANCIAL"
    case project = "PROJECT"
    case dailyLog = "DAILY_LOG"
    case custom = "CUSTOM"

    var displayName: String {
        switch self {
        case .labor: return "reports.type.labor".localized
        case .equipment: return "reports.type.equipment".localized
        case .safety: return "reports.type.safety".localized
        case .safetyMeetings: return "reports.type.safetyMeetings".localized
        case .financial: return "reports.type.financial".localized
        case .project: return "reports.type.project".localized
        case .dailyLog: return "reports.type.dailyLog".localized
        case .custom: return "reports.type.custom".localized
        }
    }

    var icon: String {
        switch self {
        case .labor: return "person.3.fill"
        case .equipment: return "wrench.and.screwdriver.fill"
        case .safety: return "shield.checkered"
        case .safetyMeetings: return "person.3.sequence.fill"
        case .financial: return "dollarsign.circle.fill"
        case .project: return "building.2.fill"
        case .dailyLog: return "doc.text.fill"
        case .custom: return "doc.badge.gearshape"
        }
    }
}

// MARK: - Report Period
enum ReportPeriod: String, Codable, CaseIterable {
    case today = "TODAY"
    case thisWeek = "THIS_WEEK"
    case thisMonth = "THIS_MONTH"
    case thisQuarter = "THIS_QUARTER"
    case thisYear = "THIS_YEAR"
    case lastWeek = "LAST_WEEK"
    case lastMonth = "LAST_MONTH"
    case lastQuarter = "LAST_QUARTER"
    case custom = "CUSTOM"

    var displayName: String {
        switch self {
        case .today: return "Today"
        case .thisWeek: return "This Week"
        case .thisMonth: return "This Month"
        case .thisQuarter: return "This Quarter"
        case .thisYear: return "This Year"
        case .lastWeek: return "Last Week"
        case .lastMonth: return "Last Month"
        case .lastQuarter: return "Last Quarter"
        case .custom: return "Custom Range"
        }
    }
}

// MARK: - Chart Type
enum ChartType: String, Codable, CaseIterable {
    case bar = "BAR"
    case line = "LINE"
    case pie = "PIE"
    case table = "TABLE"

    var displayName: String {
        switch self {
        case .bar: return "Bar Chart"
        case .line: return "Line Chart"
        case .pie: return "Pie Chart"
        case .table: return "Table"
        }
    }

    var icon: String {
        switch self {
        case .bar: return "chart.bar.fill"
        case .line: return "chart.line.uptrend.xyaxis"
        case .pie: return "chart.pie.fill"
        case .table: return "tablecells"
        }
    }
}

// MARK: - Report Summary Stat
struct ReportStat: Identifiable, Codable {
    let id: String
    let label: String
    let value: String
    let change: Double? // percentage change
    let changeLabel: String?

    var isPositiveChange: Bool {
        (change ?? 0) >= 0
    }
}

// MARK: - Report Data Point
struct ReportDataPoint: Identifiable, Codable {
    let id: String
    let label: String
    let value: Double
    let category: String?
}

// MARK: - Report
struct Report: Identifiable, Codable {
    let id: String
    let name: String
    let type: ReportType
    let description: String?
    let projectId: String?
    let projectName: String?
    let period: ReportPeriod
    let startDate: Date?
    let endDate: Date?
    let chartType: ChartType?
    let stats: [ReportStat]?
    let data: [ReportDataPoint]?
    let generatedAt: Date
    let generatedBy: String?

    // Custom init for decoding with default values
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        id = try container.decode(String.self, forKey: .id)
        name = try container.decode(String.self, forKey: .name)
        type = try container.decode(ReportType.self, forKey: .type)
        description = try container.decodeIfPresent(String.self, forKey: .description)
        projectId = try container.decodeIfPresent(String.self, forKey: .projectId)
        projectName = try container.decodeIfPresent(String.self, forKey: .projectName)
        period = try container.decode(ReportPeriod.self, forKey: .period)
        startDate = try container.decodeIfPresent(Date.self, forKey: .startDate)
        endDate = try container.decodeIfPresent(Date.self, forKey: .endDate)
        chartType = try container.decodeIfPresent(ChartType.self, forKey: .chartType)
        stats = try container.decodeIfPresent([ReportStat].self, forKey: .stats)
        data = try container.decodeIfPresent([ReportDataPoint].self, forKey: .data)
        // Default to current date if not provided
        generatedAt = try container.decodeIfPresent(Date.self, forKey: .generatedAt) ?? Date()
        generatedBy = try container.decodeIfPresent(String.self, forKey: .generatedBy)
    }

    private enum CodingKeys: String, CodingKey {
        case id, name, type, description, projectId, projectName, period
        case startDate, endDate, chartType, stats, data, generatedAt, generatedBy
    }

    // Encoding for saving to history
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(name, forKey: .name)
        try container.encode(type, forKey: .type)
        try container.encodeIfPresent(description, forKey: .description)
        try container.encodeIfPresent(projectId, forKey: .projectId)
        try container.encodeIfPresent(projectName, forKey: .projectName)
        try container.encode(period, forKey: .period)
        try container.encodeIfPresent(startDate, forKey: .startDate)
        try container.encodeIfPresent(endDate, forKey: .endDate)
        try container.encodeIfPresent(chartType, forKey: .chartType)
        try container.encodeIfPresent(stats, forKey: .stats)
        try container.encodeIfPresent(data, forKey: .data)
        try container.encode(generatedAt, forKey: .generatedAt)
        try container.encodeIfPresent(generatedBy, forKey: .generatedBy)
    }
}

// MARK: - Saved Report Template
struct SavedReport: Identifiable, Codable {
    let id: String
    let name: String
    let type: ReportType
    let description: String?
    let projectId: String?
    let filters: [String: String]?
    let columns: [String]?
    let groupBy: String?
    let sortBy: String?
    let chartType: ChartType?
    let isDefault: Bool
    let createdBy: String?
    let createdAt: Date
    let updatedAt: Date
    // Note: No explicit CodingKeys needed - APIClient uses convertFromSnakeCase

    // MARK: - Mock Data
    static let mockSavedReports: [SavedReport] = [
        SavedReport(
            id: "sr-1",
            name: "Weekly Labor Summary",
            type: .labor,
            description: "Weekly summary of labor hours by project",
            projectId: nil,
            filters: nil,
            columns: ["project", "hours", "cost"],
            groupBy: "project",
            sortBy: "hours",
            chartType: .bar,
            isDefault: true,
            createdBy: nil,
            createdAt: Date(),
            updatedAt: Date()
        ),
        SavedReport(
            id: "sr-2",
            name: "Safety Incident Trends",
            type: .safety,
            description: "Monthly safety incident trends",
            projectId: nil,
            filters: nil,
            columns: nil,
            groupBy: "month",
            sortBy: nil,
            chartType: .line,
            isDefault: false,
            createdBy: nil,
            createdAt: Date(),
            updatedAt: Date()
        )
    ]
}

// MARK: - Analytics Dashboard Data
struct AnalyticsDashboard: Codable {
    let projectMetrics: ProjectMetrics?
    let laborMetrics: LaborMetrics?
    let safetyMetrics: SafetyMetrics?
    let financialMetrics: FinancialMetrics?
    // Note: No explicit CodingKeys needed - APIClient uses convertFromSnakeCase
}

struct ProjectMetrics: Codable {
    let activeProjects: Int
    let completedProjects: Int
    let onHoldProjects: Int
    let totalBudget: Double
    let averageCompletion: Double
    // Note: No explicit CodingKeys needed - APIClient uses convertFromSnakeCase
}

struct LaborMetrics: Codable {
    let totalHoursThisWeek: Double
    let totalHoursLastWeek: Double
    let activeWorkers: Int
    let overtimeHours: Double
    let averageHoursPerWorker: Double
    // Note: No explicit CodingKeys needed - APIClient uses convertFromSnakeCase
}

struct SafetyMetrics: Codable {
    let incidentsThisMonth: Int
    let incidentsLastMonth: Int
    let daysWithoutIncident: Int
    let openInspections: Int
    let upcomingTrainings: Int
    // Note: No explicit CodingKeys needed - APIClient uses convertFromSnakeCase
}

struct FinancialMetrics: Codable {
    let totalRevenue: Double
    let totalCosts: Double
    let profitMargin: Double
    let pendingInvoices: Double
    let overdueInvoices: Double
    // Note: No explicit CodingKeys needed - APIClient uses convertFromSnakeCase
}

File: apps/android/construction-android/app/src/main/java/com/constructionpro/app/data/model/Report.kt
package com.constructionpro.app.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// Report types
object ReportType {
    const val PROJECT_SUMMARY = "PROJECT_SUMMARY"
    const val FINANCIAL = "FINANCIAL"
    const val LABOR = "LABOR"
    const val SAFETY = "SAFETY"
    const val DAILY_LOG = "DAILY_LOG"
    const val EQUIPMENT = "EQUIPMENT"
    const val WEATHER_DELAY = "WEATHER_DELAY"

    val all = listOf(DAILY_LOG, WEATHER_DELAY, PROJECT_SUMMARY, LABOR, SAFETY, FINANCIAL, EQUIPMENT)

    fun displayName(type: String): String = when (type) {
        PROJECT_SUMMARY -> "Project Summary"
        FINANCIAL -> "Financial Report"
        LABOR -> "Labor Report"
        SAFETY -> "Safety Report"
        DAILY_LOG -> "Daily Logs"
        EQUIPMENT -> "Equipment Report"
        WEATHER_DELAY -> "Weather Delays"
        else -> type.replace("_", " ")
    }
}

// Report period
object ReportPeriod {
    const val TODAY = "TODAY"
    const val THIS_WEEK = "THIS_WEEK"
    const val THIS_MONTH = "THIS_MONTH"
    const val THIS_QUARTER = "THIS_QUARTER"
    const val THIS_YEAR = "THIS_YEAR"
    const val LAST_WEEK = "LAST_WEEK"
    const val LAST_MONTH = "LAST_MONTH"
    const val LAST_QUARTER = "LAST_QUARTER"
    const val LAST_30_DAYS = "LAST_30_DAYS"
    const val LAST_90_DAYS = "LAST_90_DAYS"
    const val CUSTOM = "CUSTOM"

    val all = listOf(TODAY, THIS_WEEK, THIS_MONTH, THIS_QUARTER, THIS_YEAR, LAST_WEEK, LAST_MONTH, LAST_QUARTER, LAST_30_DAYS, LAST_90_DAYS)

    fun displayName(period: String): String = when (period) {
        TODAY -> "Today"
        THIS_WEEK -> "This Week"
        THIS_MONTH -> "This Month"
        THIS_QUARTER -> "This Quarter"
        THIS_YEAR -> "This Year"
        LAST_WEEK -> "Last Week"
        LAST_MONTH -> "Last Month"
        LAST_QUARTER -> "Last Quarter"
        LAST_30_DAYS -> "Last 30 Days"
        LAST_90_DAYS -> "Last 90 Days"
        CUSTOM -> "Custom Range"
        else -> period
    }
}

@Serializable
data class ReportOverview(
    val activeProjects: Int? = null,
    val timeEntries: Int? = null,
    val dailyLogs: Int? = null,
    val incidents: Int? = null,
    val equipment: Int? = null,
    val totalHours: Double? = null,
    val pendingApprovals: Int? = null,
    val openIssues: Int? = null,
    val completedTasks: Int? = null
)

// ============ ANALYTICS MODELS ============

@Serializable
data class AnalyticsDashboard(
    val summary: AnalyticsSummary? = null,
    val projectMetrics: List<ProjectMetric>? = null,
    val laborMetrics: LaborMetrics? = null,
    val safetyMetrics: SafetyMetrics? = null,
    val financialMetrics: FinancialMetrics? = null,
    val trends: TrendData? = null
)

@Serializable
data class AnalyticsSummary(
    val totalProjects: Int = 0,
    val activeProjects: Int = 0,
    val completedProjects: Int = 0,
    val onHoldProjects: Int = 0,
    val totalBudget: Double = 0.0,
    val totalSpent: Double = 0.0,
    val totalHoursLogged: Double = 0.0,
    val avgProjectCompletion: Double = 0.0
)

@Serializable
data class ProjectMetric(
    val projectId: String,
    val projectName: String,
    val status: String,
    val progress: Double = 0.0,
    val budget: Double = 0.0,
    val spent: Double = 0.0,
    val hoursLogged: Double = 0.0,
    val dailyLogCount: Int = 0,
    val openIssues: Int = 0,
    val startDate: String? = null,
    val endDate: String? = null
)

@Serializable
data class LaborMetrics(
    val totalHours: Double = 0.0,
    val regularHours: Double = 0.0,
    val overtimeHours: Double = 0.0,
    val totalWorkers: Int = 0,
    val avgHoursPerWorker: Double = 0.0,
    val hoursThisWeek: Double = 0.0,
    val hoursLastWeek: Double = 0.0,
    val weeklyChange: Double = 0.0
)

@Serializable
data class SafetyMetrics(
    val totalIncidents: Int = 0,
    val openIncidents: Int = 0,
    val closedIncidents: Int = 0,
    val daysWithoutIncident: Int = 0,
    val inspectionsDue: Int = 0,
    val inspectionsCompleted: Int = 0,
    val safetyScore: Double = 0.0
)

@Serializable
data class FinancialMetrics(
    val totalBudget: Double = 0.0,
    val totalSpent: Double = 0.0,
    val remainingBudget: Double = 0.0,
    val budgetUtilization: Double = 0.0,
    val invoicesPending: Int = 0,
    val invoicesOverdue: Int = 0,
    val totalInvoiced: Double = 0.0,
    val totalPaid: Double = 0.0
)

@Serializable
data class TrendData(
    val dailyLogTrend: List<TrendPoint>? = null,
    val hoursTrend: List<TrendPoint>? = null,
    val incidentTrend: List<TrendPoint>? = null,
    val spendingTrend: List<TrendPoint>? = null
)

@Serializable
data class TrendPoint(
    val date: String,
    val value: Double,
    val label: String? = null
)

// ============ REPORT GENERATION ============

@Serializable
data class ReportRequest(
    val type: String,
    val projectId: String? = null,
    val period: String = ReportPeriod.THIS_MONTH,
    val startDate: String? = null,
    val endDate: String? = null,
    val format: String = "PDF" // PDF, CSV, XLSX
)

// Report stat for generated reports
@Serializable
data class ReportStat(
    val id: String,
    val label: String,
    val value: String,
    val change: Double? = null,
    @SerialName("change_label")
    val changeLabel: String? = null
)

// Report data point for chart data
@Serializable
data class ReportDataPoint(
    val id: String,
    val label: String,
    val value: Double,
    val category: String? = null
)

@Serializable
data class GeneratedReport(
    val id: String,
    val type: String,
    val name: String,
    val status: String, // GENERATING, READY, FAILED
    val format: String,
    val description: String? = null,
    val fileUrl: String? = null,
    val chartType: String? = null,
    val stats: List<ReportStat>? = null,
    val data: List<ReportDataPoint>? = null,
    val createdAt: String,
    val projectId: String? = null,
    val projectName: String? = null,
    val period: String? = null,
    val startDate: String? = null,
    val endDate: String? = null,
    val generatedBy: String? = null
)

@Serializable
data class ReportListResponse(
    val reports: List<GeneratedReport>
)

// ============ PROJECT REPORT DETAILS ============

@Serializable
data class ProjectReport(
    val project: ProjectReportSummary,
    val dailyLogs: DailyLogReportSection? = null,
    val labor: LaborReportSection? = null,
    val materials: MaterialsReportSection? = null,
    val issues: IssuesReportSection? = null,
    val photos: Int = 0
)

@Serializable
data class ProjectReportSummary(
    val id: String,
    val name: String,
    val status: String,
    val client: String? = null,
    val address: String? = null,
    val startDate: String? = null,
    val endDate: String? = null,
    val progress: Double = 0.0,
    val budget: Double = 0.0,
    val spent: Double = 0.0
)

@Serializable
data class DailyLogReportSection(
    val total: Int = 0,
    val approved: Int = 0,
    val pending: Int = 0,
    val avgCrewSize: Double = 0.0,
    val totalHours: Double = 0.0
)

@Serializable
data class LaborReportSection(
    val totalHours: Double = 0.0,
    val regularHours: Double = 0.0,
    val overtimeHours: Double = 0.0,
    val uniqueWorkers: Int = 0,
    val laborCost: Double = 0.0
)

@Serializable
data class MaterialsReportSection(
    val totalItems: Int = 0,
    val totalCost: Double = 0.0,
    val delivered: Int = 0,
    val pending: Int = 0
)

@Serializable
data class IssuesReportSection(
    val total: Int = 0,
    val open: Int = 0,
    val resolved: Int = 0,
    val critical: Int = 0
)
